### PR TITLE
Feature/pbdw

### PR DESCRIPTION
**Pull request by romainhild**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [ ] Have you successfully run the Feel++ testsuite with your changes locally?
- [x] Have you written Doxygen comments in your contribution ?

-----

- add PbdwOnline class without template for online phase of PBDW
- add possibility to use only some sensors during online phase
- add outputs to PBDW
- add outputs to toolboxmor
- add method to initialize toolboxmor directly with a toolbox
- add python wrapping for PBDW
- add test for PBDW

### CODE REVIEW

File: feelpp/feel/feeldiscr/sensors.hpp [hunk 1]
```diff
@@ -240,9 +240,14 @@ class SensorMap : public std::map<std::string, std::shared_ptr<SensorBase<Space>
...
     explicit SensorMap( space_ptrtype const& Xh ): M_Xh(Xh) {}
     explicit SensorMap( space_ptrtype const& Xh, json j)
-        : M_Xh(Xh),
-          M_j(j)
+        : M_Xh(Xh)
     {
+        this->loadJson(j);
+    }
+
+    void loadJson(json j)
```
Comment (prudhomm) [inline feelpp/feel/feeldiscr/sensors.hpp:248]: `json const& j ` otherwise copy all the time

File: feelpp/feel/feelmodels/modeloutputs.cpp [hunk 2]
```diff
@@ -52,16 +52,25 @@ ModelOutput::ModelOutput( std::string name, pt::ptree const& p, worldcomm_ptr_t
     if( auto markers = p.get_child_optional("markers") )
         M_markers.setPTree(*markers);
     else
-        LOG(WARNING) << "Output " << M_name << " does not have any range\n";
+        LOG(INFO) << "Output " << M_name << " does not have any range\n";
 
     if( auto t = p.get_optional<int>("topodim") )
```
Comment (prudhomm) [inline feelpp/feel/feelmodels/modeloutputs.cpp:57]: it seems we are using boost property_tree, why is that ? I thought we moved away, no?

File: mor/mor/feel/feelmor/pbdw.hpp [hunk 3]
```diff
@@ -38,15 +38,326 @@
...
+    }
+}
+
 /**
  * Class that implements Parameterized Background Data-Weak method.
  * @tparam Reduced basis space
  */
 template<typename RBSpace>
-class PBDW : CRBDB
+class PBDW : public PBDWOnline
```
Comment (prudhomm) [inline mor/mor/feel/feelmor/pbdw.hpp:357]: why depend on PBDWOnline ? why not a member instead ?

File: mor/mor/feel/feelmor/toolboxmor.hpp [hunk 4]
```diff
@@ -17,6 +17,65 @@ makeToolboxMorOptions();
 FEELPP_EXPORT AboutData
 makeToolboxMorAbout( std::string const& str = "opusheat-tb" );
 
+template<typename MeshType>
+class ToolboxMorModelBase
```
Comment (prudhomm) [inline mor/mor/feel/feelmor/toolboxmor.hpp:21]: add documentation for the class
use doxygen mode from vscode to help with this

File: mor/mor/feel/feelmor/toolboxmor_impl.hpp [hunk 5]
```diff
@@ -271,27 +409,76 @@ ToolboxMor<SpaceType, Options>::assembleData()
...
+            }
+            else if( dim == nDim-1 )
+            {
+                auto range = markedfaces(this->Xh->mesh(), out.markers());
+                double area = integrate( _range=range, _expr=cst(1.0) ).evaluate()(0,0) ;
+                fAvg += integrate( _range=range, _expr=id(u)/cst(area) );
+            }
+            this->M_Fqm[output++][0][0] = fAvg.vectorPtr();
+        }
+        else if( out.type() == "flux")
```
Comment (prudhomm) [inline mor/mor/feel/feelmor/toolboxmor_impl.hpp:438]: this is the normal_derivative, not the flux
the flux would have the conduction coefficient -k*grad(u)*N

File: mor/pyfeelpp-mor/feelpp/mor/__init__.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import feelpp
 from ._mor import *
 from ._toolboxmor import *
+from ._pbdw import *
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/feelpp/mor/__init__.py:4]: may be we should have pbdw within its own separate module instead
it should probably not be loaded by default when we load pyfeelpp-mor

File: mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        .def("online", py::overload_cast<vectorN_type const&, std::vector<int> const&, bool>(&pbdw_t::online, py::const_), "returns the coefficients of the solution", py::arg("yobs"), py::arg("sensors"), py::arg("toComplete")=true )
+        .def("outputs", py::overload_cast<vectorN_type const&, std::vector<int> const&, bool>(&pbdw_t::outputs, py::const_), "returns the outputs of the mode", py::arg("yobs"), py::arg("sensors"), py::arg("toComplete")=true )
+        .def("online", py::overload_cast<vectorN_type const&, std::vector<std::string> const&>(&pbdw_t::online, py::const_), "returns the coefficients of the solution", py::arg("yobs"), py::arg("sensors") )
+        .def("outputs", py::overload_cast<vectorN_type const&, std::vector<std::string> const&>(&pbdw_t::outputs, py::const_), "returns the outputs of the mode", py::arg("yobs"), py::arg("sensors") )
+        .def("onlineWithout", py::overload_cast<vectorN_type const&, std::vector<int> const&>(&pbdw_t::onlineWithout, py::const_), "returns the coefficients of the solution", py::arg("yobs"), py::arg("sensors") )
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp:55]: Without what ?
the name is not very expressive

Comment (prudhomm) [reply mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp:55]: also the documentation needs to be improved

File: mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        .def("outputsWithout", py::overload_cast<vectorN_type const&, std::vector<int> const&>(&pbdw_t::outputsWithout, py::const_), "returns the outputs of the mode", py::arg("yobs"), py::arg("sensors") )
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp:56]: idem here and below

**Review by prudhomm — COMMENTED** [review]
We need to discuss PBDW inheritance with PBDWOnline

File: feelpp/feel/feelmodels/modeloutputs.cpp (hunk 2 above)
Comment (romainhild) [reply feelpp/feel/feelmodels/modeloutputs.cpp:57]: Not yet, I think @vincentchabannes is doing it in the branch feature/toolboxes

**Comment by romainhild** [comment]
> @romainhild there are a lot of compilation issues in this PR https://buildkite.com/feelpp/feelpp/builds/4188#8cb7b42f-ffb4-4729-b494-0b61a1f438d3/265-521

I'll look into it, it seems something is wrong in 2D.

File: mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp (hunk 7 above)
Comment (romainhild) [reply mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp:55]: It is without the list of sensors given in argument. 
But I don't really use it in fact, maybe we can remove those methods and use only the one where we give the list of sensors used.

File: mor/mor/feel/feelmor/pbdw.hpp (hunk 3 above)
Comment (romainhild) [reply mor/mor/feel/feelmor/pbdw.hpp:357]: I don't see why PBDW should own PBDWOnline. It is just a layer that add capability to do the offline phase.

File: mor/mor/feel/feelmor/toolboxmor.hpp (hunk 4 above)
Comment (romainhild) [reply mor/mor/feel/feelmor/toolboxmor.hpp:21]: Ok, but I think we should also rename ToolboxMor to something like DEIMMor or something. It is not really related to the toolboxes, but more to DEIM, we could use it without toolboxes.
What do you think ?

**Comment by romainhild** [comment]
@prudhomm @vincentchabannes I think this is ready to be merged.
Can you review the changes ?

File: feelpp/feel/feelmodels/modeloutputs.hpp [hunk 9]
```diff
@@ -58,6 +58,8 @@ class FEELPP_EXPORT ModelOutput : public CommObject
     std::string name() const { return M_name; }
     std::string type() const { return M_type; }
     std::set<std::string> markers() const { return M_markers; }
+    std::vector<double> coord() const { return M_coord; }
```
Comment (prudhomm) [inline feelpp/feel/feelmodels/modeloutputs.hpp:61]: std::vector<double>. const&

File: feelpp/pyfeelpp/feelpp/operators.cpp [hunk 10]
```diff
@@ -72,6 +73,16 @@ void defOperator( py::module& m )
             return a.matrixPtr(); },
         py::arg( "trial" ), py::arg( "test" ), py::arg( "range" ), py::arg( "coeff" )=std::string{"1"}, "create a mass matrix" );
     m.def(
+        "mass", []( space_ptr_t const& domain, space_ptr_t const& image, faces_iterator_range_t const& r, std::string const& c = "1" )
```
Comment (prudhomm) [inline feelpp/pyfeelpp/feelpp/operators.cpp:76]: the code is redundant with element wise mass lambda function.
this needs to be made generic wrt elements and faces ranges

File: mor/pyfeelpp-mor/feelpp/mor/__init__.py (hunk 6 above)
Comment (prudhomm) [reply mor/pyfeelpp-mor/feelpp/mor/__init__.py:4]: should be in separate module

File: mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+PYBIND11_MODULE(_pbdw, m )
+{
+    using namespace Feel;
+    using pbdw_t = PBDWOnline;
+    using vectorN_type = typename pbdw_t::vectorN_type;
+    using matrixN_type = typename pbdw_t::matrixN_type;
+    py::class_<pbdw_t,std::shared_ptr<pbdw_t>>(m, "Pbdw")
+        .def(py::init<std::string const&, int, std::string const&, std::string const&>(),
+             py::arg("name")=std::string("pbdw"),
+             py::arg("dbLoad")=1,
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp:37]: should be an enum no ?

File: mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py [hunk 12]
```diff
@@ -0,0 +1,895 @@
+"""reducedBasisPetsc :
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py:1]: you merged pyrb into feature.pbdw ?

File: mor/pyfeelpp-mor/tests/CMakeLists.txt [hunk 13]
```diff
@@ -1,22 +1,22 @@
...
+set(PYFILES conftest.py test_mor.py)
 install(FILES ${PYFILES} DESTINATION ${FEELPP_PYTHON_MODULE_PATH}/feelpp)
 
 foreach(test mor )
   add_test (NAME feelpp_mor_python-tests-${test}
     COMMAND ${CMAKE_COMMAND} -E env
           LD_LIBRARY_PATH=${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}:$ENV{LD_LIBRARY_PATH}
           PYTHONPATH=${CMAKE_INSTALL_PREFIX}/${FEELPP_PYTHON_MODULE_PATH}
-          ${PYTHON_EXECUTABLE} -m pytest -k ${test} -s ${CMAKE_CURRENT_SOURCE_DIR}
+          ${PYTHON_EXECUTABLE} -m pytest test_${test}.py -s -vv
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/CMakeLists.txt:12]: why do that ? it was more general before ?

File: mor/pyfeelpp-mor/tests/CMakeLists.txt [hunk 14]
```diff
@@ -1,22 +1,22 @@
...
           WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
   )
   add_test (NAME feelpp_mor_python-tests-mpi-${test}
     COMMAND ${CMAKE_COMMAND} -E env
           LD_LIBRARY_PATH=${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}:$ENV{LD_LIBRARY_PATH}
           PYTHONPATH=${CMAKE_INSTALL_PREFIX}/${FEELPP_PYTHON_MODULE_PATH}
-          mpirun -np 2 ${PYTHON_EXECUTABLE} -m pytest -k ${test} -s ${CMAKE_CURRENT_SOURCE_DIR}
+          mpirun -np 2 ${PYTHON_EXECUTABLE} -m pytest test_${test}.py -s -vv
```
Comment (prudhomm) [inline mor/pyfeelpp-mor/tests/CMakeLists.txt:19]: idem

**Review by prudhomm — CHANGES_REQUESTED** [review]

File: mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp (hunk 11 above)
Comment (romainhild) [reply mor/pyfeelpp-mor/feelpp/mor/pbdw.cpp:37]: this is the general behavior of the option in crb, I didn't want to change everything, so I used the same thing.

File: mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py (hunk 12 above)
Comment (romainhild) [reply mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py:1]: yes, I needed some of the things that @thomas-saigre did in this branch, and it was in several commits.
The branch is ready to be merge also in develop no ?

File: mor/pyfeelpp-mor/tests/CMakeLists.txt (hunk 13 above)
Comment (romainhild) [reply mor/pyfeelpp-mor/tests/CMakeLists.txt:12]: this is from pyrb #1787

File: mor/pyfeelpp-mor/tests/CMakeLists.txt (hunk 14 above)
Comment (prudhomm) [reply mor/pyfeelpp-mor/tests/CMakeLists.txt:19]: ok it is from branch pyrb.

File: feelpp/pyfeelpp/feelpp/operators.cpp (hunk 10 above)
Comment (romainhild) [reply feelpp/pyfeelpp/feelpp/operators.cpp:76]: this is from pyrb #1787

**Comment by prudhomm** [comment]
@romainhild could you fix the little things above ?

File: mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py (hunk 12 above)
Comment (thomas-saigre) [reply mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py:1]: not yet... there are still issues with the error bound, I'm working on it right now, and the parabolic code is not working at all

Comment (prudhomm) [reply mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py:1]: I think it is fine as long as the code and the tests are fine. even if everything is not ready

Comment (prudhomm) [reply mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py:1]: ready meaning coded

File: mor/pyfeelpp-mor/tests/CMakeLists.txt (hunk 13 above)
Comment (thomas-saigre) [reply mor/pyfeelpp-mor/tests/CMakeLists.txt:12]: it was not working the way it was before, tests were not launched at all... But maybe that can be changed

File: mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py (hunk 12 above)
Comment (thomas-saigre) [reply mor/pyfeelpp-mor/feelpp/mor/reducedbasis/reducedbasis.py:1]: I try to commit the fixes this morning, if it is ok ?

**Comment by prudhomm** [comment]
@romainhild empirical quadrature fails: https://buildkite.com/feelpp/feelpp/builds/4243#9edef1ff-364c-4938-b426-742839d1338b/759-815

**Comment by romainhild** [comment]
@prudhomm this is a test with a random set of parameters, it seems that for this set we did not reach the tolerance (error: in "eq_suite/test_1": absolute value of i11 - i12{1.0829318330962678e-08} exceeds 1e-08).
I'll use a greater tolerance.

**Pull request merged**